pkg/maps/vtep/vtep.go

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package vtep
+
+import (
+	"fmt"
+	"net"
+	"unsafe"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/types"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "map-vtep")
+
+const (
+	// MaxEntries is the maximum number of keys that can be present in the
+	// VTEP map.
+	MaxEntries = defaults.MaxVTEPDevices
+
+	// Name is the canonical name for the VTEP map on the filesystem.
+	Name = "cilium_vtep_map"
+)
+
+// Key implements the bpf.MapKey interface.
+//
+// Must be in sync with struct vtep_key in <bpf/lib/common.h>
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type Key struct {
+	IP types.IPv4 `align:"vtep_ip"`
+}
+
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k *Key) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// NewValue returns a new empty instance of the structure representing the BPF
+// map value
+func (k Key) NewValue() bpf.MapValue { return &VtepEndpointInfo{} }
+
+func (k Key) String() string {
+	ipStr := net.IP(k.IP[:net.IPv4len]).String()
+	return fmt.Sprintf("%s", ipStr)
+}
+
+// NewKey returns an Key based on the provided IP address and mask.
+func NewKey(ip net.IP) Key {
+	result := Key{}
+
+	if ip4 := ip.To4(); ip4 != nil {
+		copy(result.IP[:], ip4)
+	}
```

Comment:
The datapath takes a packet destination, masks it with the VTEP CIDR mask, then performs the lookup. However, this key creation & insert from the control plane directly takes the user input and passes it into the map. I believe this means that if the user specifies `--vtep-cidr=192.168.0.1 --vtep-mask=255.255.255.0` then the lookup will fail, because the map is populated with `192.168.0.1` instead of `192.168.0.0` (ie the masked version). This is a minor usability nit but could catch someone out, I think. Maybe this `NewKey()` function should also apply the mask against the CIDRs when creating the keys, to ensure that they are properly formatted before inserting into the datapath map?

Reply:
sorry be to rough on golang here :), I did this `ip4 = ip4 & option.Config.VtepCidrMask` and I got error

```
invalid operation: operator & not defined on ip4 (variable of type net.IP)
```
tried google search proper golang bitwise mask operation on IP, no result

Reply:
I would suggest using the proper IPMask type to apply the masking: https://pkg.go.dev/net#IP.Mask

Reply:
silly me :) there is already Mask function 
```
      if ip4 := ip.To4(); ip4 != nil {
+               ip4.Mask(net.IPMask(option.Config.VtepCidrMask))
                copy(result.IP[:], ip4)
        }
```